### DISCUSSION TITLE
Popover component won't always close when clicking outside of it

### DISCUSSION THREAD

**Discussion by FabioDainese** [tailwindlabs/headlessui#2731] [Help]
**What package within Headless UI are you using?**
_@headlessui/react_  (more specifically the ***popover*** component)

**What version of that package are you using?**
_v1.7.17_

**What browser are you using?**
_Chrome (v116.0.5845.96), Firefox (v116.0.3), Safari (v16.4.1)_

**Reproduction URL**
[https://codesandbox.io/p/sandbox/flamboyant-marco-57cyzh](https://codesandbox.io/p/sandbox/flamboyant-marco-57cyzh)

**Describe your issue**
The *popover* component seems to not close every time you click outside of it. To be more specific - you can try on the provided sandbox - only when you click on one of the siblings of the component that uses the popover, it won't work. So, for example, in the provided project if you click on:

* the **green div**: the popover will still be open (i.e. the `<main>` tag is sibling to the `<nav>` one, which uses the *popover* component)
* the **white div**: the popover will close (i.e. the `<body>` tag is parent of the `<nav>`, so all okay)
* the "**`'body' tag - white background`" text**: the popover will still be open (i.e. the `<p>` tag that contains the text is sibling to the `<nav>` one)

Now, the temporary "*workaround*" that kinda fix this issue would be to use a transparent overlay (i.e. `<Popover.Overlay className="fixed inset-0" />`), so basically wherever you click, you would click on the overlay that consequently will close the popover. Now, this is far from optimal though, since if a user would like to click on an hypothetical visible button (placed outside the popover), it would need to click 2 times (one to close the popover by clicking the invisible overlay and the second one to actually trigger the button).

So, I'm not sure if this is the intended behavior for the `popover` component - or if I coded something wrong here - but I would like to gently ask you a feedback on it, also to understand which would be the best approach to tackle this issue (if it's not the intended behavior).

P.S. the popover used on the provided project is just a simpler copy-paste version of the first example that you can find on the [headlessui website](https://headlessui.com/react/popover)

**Comment by mattmcegg** [comment]
Seeing this behavior on v1.7.17, but working as expected on v1.7.14

**Reply by FabioDainese** [reply]
Thank you so much @mattmcegg for this discovery 🙌 much appreciated

Question, should I transform this discussion into an *issue*, just to mark it as an "official" bug (and keep track of its development)?

**Reply by madmaxmehdi44** [reply]
### Following the error that occurs while clicking, I will also introduce you an npm package that will make the work easier. 
> [!NOTE]
> Although it is possible that most of the readers and friends are fully aware of this topic by now.

**`$npm install use-onclickoutside`**

### How to Usage???

`import useOnClickOutside from "use-onclickoutside";`
`const searchRef = React.useRef(null);`


```
useOnClickOutside(searchRef, () => {
     setQuery("");
  });     

```
```
const onSearch = (e: { target: { value: any } }) => {
    const value = e.target.value;
    setQuery(value);
 };            
```

**Comment by ijrzhong** [comment]
Tested it and found the bug existed on latest 3 versions for headlessui/react: 1.7.17, 1.7.16 and 1.7.15, but not on 1.7.14 as @mattmcegg said. Mitigate it temporarily by using 1.7.14 and you could create an issue for a fix.
It seems that ancestors of the popover component work as expected, but not for the case in children of its ancestors. The bug exists for keyboard interaction as well.

**Reply by markdost** [reply]
Same issue. Downgrading to `"@headlessui/react": "1.7.14",` solves the issue for now.

**Comment by tomelliot** [comment]
I also had this issue with `1.7.17`. Downgrading to `1.7.14` has resolved it for now.

**Comment by AngelaRucci** [comment]
I am having this issue still on 1.7.14

**Reply by FabioDainese** [reply]
Have you tried [this solution/workaround](https://github.com/tailwindlabs/headlessui/issues/2752#issuecomment-1724096430)?

**Reply by laolee010126** [reply]
@FabioDainese thanks a lot! it worked for me (using headless ui with nextjs app router)

**Comment by halanizi** [comment]
This issue still exists on version 1.7.18 as well.

**Comment by ermakovich** [comment]
The issue exists on latest version (2.2.2) of `@headlessui/react` as well.

**Reply by grassroots-labs-admin** [reply]
Seeing this primarily on mobile and curious if I set something up incorrectly, and the popover doesn't close when I click externally to the dropdown menu:
```
      <div className={cn("w-full flex flex-wrap items-center gap-2", className)}>
            <Popover className="relative">
                {({ open, close }) => (
                    <>
                        <PopoverButton as={IconButton} // make IconButton render a <button>
                            buttonType="button"
                            svgIcon={<PlusCircleIcon size={24} />}
                            className={cn(
                                'h-8 px-2 rounded-full',
                                'bg-gray-100 hover:bg-gray-200'
                            )}
                            ariaLabel="add-new-tab-button"
                            onClick={() => {}}  
                            />

                        <PopoverPanel 
                            className="absolute left-full top-0 z-[999] ml-2 min-w-[300px] 
                                                flex flex-col gap-2 rounded-xl bg-white shadow-lg 
                                                 ring-1 ring-black ring-opacity-5 max-h-[200px] 
                                                overflow-y-auto overflow-x-visible">
                            <div className="w-full flex flex-wrap gap-2" aria-label="choose-available-tab-popover">
                                {sortedOptions
                                    .filter(option => !selectedIds.includes(option.id))
                                    .map(option => (
                                        <div 
                                            key={option.id} 
                                            onClick={() => {
                                                toggleTab(option.id);
                                            }}
                                            className='py-[5px]'
                                            aria-label="available-tab-option"
                                        >
                                            {renderTab(option)}
                                        </div>
                                    ))}
                            </div>
                        </PopoverPanel>
                    </>
                )}
            </Popover>
            </div>
 ```
